### PR TITLE
Union code coverage

### PR DESCRIPTION
**Pull request by kddubey**
#128 reduced code coverage b/c the current coverage workflow runs the tests w/o `pydantic` installed.

We want the union of lines covered by the 3 testing environments in the testing workflow: one without `pydantic` installed, one with `pydantic<2`, and one with `pydantic>=2`. This funcionality is achieved through [`pytest-cov`'s `--cov-append` feature](https://pytest-cov.readthedocs.io/en/latest/config.html).

## How has this been tested?

Locally. Create 3 virtual environments: one without `pydantic` installed, one with `pydantic<2`, and one with `pydantic>=2`. Then simulate the test workflow:

```zsh
workon tap-wo-pydantic && \
pytest --cov=tap && \
workon tap-pydantic-v1 && \
pytest --cov=tap --cov-append && \
workon tap && \
pytest --cov=tap --cov-append
```

This command outputs 3 coverage reports (one for each `pytest` invocation):

```
# VIRTUAL ENVIRONMENT: tap-wo-pydantic
# test output omitted

---------- coverage: platform darwin, python 3.11.8-final-0 ----------
Name              Stmts   Miss  Cover
-------------------------------------
tap/__init__.py       5      0   100%
tap/_version.py       3      0   100%
tap/tap.py          290     21    93%
tap/tapify.py       156     22    86%
tap/utils.py        230     22    90%
-------------------------------------
TOTAL               684     65    90%


# VIRTUAL ENVIRONMENT: tap-pydantic-v1
# test output omitted

---------- coverage: platform darwin, python 3.11.8-final-0 ----------
Name              Stmts   Miss  Cover
-------------------------------------
tap/__init__.py       5      0   100%
tap/_version.py       3      0   100%
tap/tap.py          290     21    93%
tap/tapify.py       156     14    91%
tap/utils.py        230     22    90%
-------------------------------------
TOTAL               684     57    92%


# VIRTUAL ENVIRONMENT: tap
# test output omitted

---------- coverage: platform darwin, python 3.11.8-final-0 ----------
Name              Stmts   Miss  Cover
-------------------------------------
tap/__init__.py       5      0   100%
tap/_version.py       3      0   100%
tap/tap.py          290     21    93%
tap/tapify.py       156      3    98%
tap/utils.py        230     22    90%
-------------------------------------
TOTAL               684     46    93%

```

The final `.coverage` file that should be sent to codecov.io is that last one (obtained locally via `coverage report -m`):

```
Name              Stmts   Miss  Cover   Missing
-----------------------------------------------
tap/__init__.py       5      0   100%
tap/_version.py       3      0   100%
tap/tap.py          290     21    93%   210, 239-241, 402-404, 438, 467, 514, 558, 561-566, 587, 686-687, 717, 725
tap/tapify.py       156      3    98%   89, 143, 165
tap/utils.py        230     22    90%   83-85, 147, 279, 305, 426, 439-455, 487-491
-----------------------------------------------
TOTAL               684     46    93%
```

So this PR is correct if and only if the coverage seen in codecov.io matches the output immediately above. Expect coverage to be: (684-46)/684 = 93.27%.

## Explanation of code coverage

Recall that #128 only touches `tap/__init__.py` and `tap/tapify.py`. 2 of the 3 uncovered lines in `tap/tapify.py` are inconsequential:
* [89](https://github.com/swansonk14/typed-argument-parser/blob/8147f75eb821f4b592067f0cb8431af9b446b90c/tap/tapify.py#L89) is moot b/c of [242-244](https://github.com/swansonk14/typed-argument-parser/blob/8147f75eb821f4b592067f0cb8431af9b446b90c/tap/tapify.py#L242-L244)
* [143](https://github.com/swansonk14/typed-argument-parser/blob/8147f75eb821f4b592067f0cb8431af9b446b90c/tap/tapify.py#L143) raises a `TypeError` that can't be hit (unless a curious user manually applies `_tap_data_from_data_model` to a non-data-model) b/c of [245](https://github.com/swansonk14/typed-argument-parser/blob/8147f75eb821f4b592067f0cb8431af9b446b90c/tap/tapify.py#L245)

[165](https://github.com/swansonk14/typed-argument-parser/blob/8147f75eb821f4b592067f0cb8431af9b446b90c/tap/tapify.py#L165) raises a `TypeError` that can be hit if a user inputs a data model containing a field that isn't a Pydantic model field, Pydantic dataclass field, or builtin dataclass field. This should be tested later, but hopefully it's ok for now :-)

### CODE REVIEW

**Comment by kddubey** [comment]
> So this PR is correct if and only if the coverage seen in codecov.io matches the output immediately above. Expect coverage to be: (684-46)/684 = 93.27%.

Coverage turned out to be 94% but I'm not totally sure why lol

Edit: I think it's b/c there are Python version-dependent clauses that my local test didn't hit

File: README.md [hunk 1]
```diff
@@ -81,6 +105,10 @@ pip install -e .
   + [tapify help](#tapify-help)
   + [Command line vs explicit arguments](#command-line-vs-explicit-arguments)
   + [Known args](#known-args)
+* [Convert to a `Tap` class](#convert-to-a-tap-class)
```
Comment (kddubey) [inline README.md:108]: Forgot to link the added documentation from #128 in the table of contents

**Comment by martinjm97** [comment]
@kddubey , this looks excellent to me! Thank you for adding the missing link too. As long as @swansonk14 approves, it should be ready for release.

--Jesse

File: .github/workflows/tests.yml [hunk 2]
```diff
@@ -45,12 +44,18 @@ jobs:
...
         python -m pip install "pydantic < 2"
-        pytest
+        pytest --cov=tap --cov-append
     - name: Test with pydantic v2
       run: |
         python -m pip install "pydantic >= 2"
-        pytest
+        pytest --cov=tap --cov-append
+
+    - name: Upload coverage reports to Codecov
```
Comment (kddubey) [inline .github/workflows/tests.yml:57]: FYI new behavior (evidently) runs a code coverage report for PRs to main. Current behavior only runs a coverage report on pushes to main

File: README.md [hunk 3]
```diff
@@ -42,18 +42,42 @@ Running `python square.py --num 2` will print `The square of your number is 4.0.
...
+
+<details>
+<summary>To develop this package, install development requirements (in a virtual environment):</summary>
+
+```
+python -m pip install -e ".[dev]"
+```
+
+Style:
+- Please use [`black`](https://github.com/psf/black) formatting
```
Comment (kddubey) [inline README.md:69]: Pretty sure this is [accurate](https://github.com/swansonk14/typed-argument-parser/commit/52df386df01a4bb9e060d366509fed470da55160)

File: .github/workflows/tests.yml (hunk 2 above)
Comment (kddubey) [reply .github/workflows/tests.yml:57]: Maybe [there's a way](https://stackoverflow.com/a/64222104/18758987) to upload the final coverage report after the matrix is done. Lmk if that behavior is more desirable than the current one, which uploads a coverage report per matrix entry (15 of them)

Comment (martinjm97) [reply .github/workflows/tests.yml:57]: That sounds like an improvement to me!

**Comment by swansonk14** [comment]
This looks awesome, thank you @kddubey!

**Pull request merged**